keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md

```diff
@@ -99,10 +107,27 @@ The kube-controller-manager will still import the cloud provider implementations
 
 #### Phase 3 - Migrating Provider Code to Provider Repos
 
-In Phase 3, all code in `k8s.io/kubernetes/staging/src/k8s.io/legacy-cloud-providers/<provider>` will be removed and development of each cloud provider should be done in their respective external repos. It's important that by this phase, both in-tree and out-of-tree cloud providers are tested and production ready. Ideally most Kubernetes clusters in production should be using the out-of-tree provider before in-tree support is removed. A plan to migrate existing clusters from using the `kube-controller-manager` to the `cloud-controller-manager` is currently being developed. More details soon.
+In Phase 3, feature development is no longer accepted in `k8s.io/kubernetes/staging/src/k8s.io/legacy-cloud-providers/<provider>` and development of each cloud provider should be done in their respective external repos. Only bug and security fixes are accepted in-tree during this phase. It's important that by this phase, both in-tree and out-of-tree cloud providers are tested and production ready. Ideally most Kubernetes clusters in production should be using the out-of-tree provider before in-tree support is removed. A plan to migrate existing clusters from using the `kube-controller-manager` to the `cloud-controller-manager` is currently being developed. More details soon.
 
 External cloud providers can optionally still import providers from `k8s.io/legacy-cloud-providers` but no core components in `k8s.io/kubernetes` will import the legacy provider and the respective staging directory will be removed along with all its dependencies.
 
+#### Phase 4 - Disabling In-Tree Providers
```

Comment:
How does these feature gates interact with `InTreePlugin<cloud>Unregister`, such as `InTreePluginGCEUnregister`? They should be enabled / disabled and graduate in sync. Does it make sense to have two flags for it? At least interaction with those feature gates should be described in this KEP.

Reply:
I didn't call specific providers, but see line 120 on CSI migration

Reply:
I believe it does make sense. @jsafrane, I believe the flags you referenced are specifically for controlling the CSI plugins for each cloud provider, which do not rely on either the cloud provider interface or any of its implementations. @andrewsykim is adding controls for the cloud provider implementations and the credential provider plugin ( which is separate from either CSI or cloud provider). Having these three sets if flags allows us to migrate out the various in-tree cloud providers, without having to keep all three sections in lock-step.

Reply:
Let me rephrase my concern, using GCE as an example. This proposal says "*this feature gate will disable any functionality in kube-apiserver, kube-controller-manager and kubelet related to the `--cloud-provider` component flag*". To me it seems that when the feature is enabled, cloud providers are removed from kube-controller-manager.

But attach-detach controller in kube-controller-manager may still use in-tree GCE volume plugin, which in turn uses in-tree GCE cloud provider to attach/detach volumes and it will fail if the cloud provider is not there.

Usage of in-tree GCE volume plugin is controlled by `InTreePluginGCEUnregister`, which won't be on by default until CSI migration for GCE is GA (~1.23?, other plugins even later).